crates/ty_server/src/server/main_loop.rs

```diff
@@ -25,19 +26,48 @@ impl Server {
 
             match next_event {
                 Event::Message(msg) => {
-                    if self.connection.handle_shutdown(&msg)? {
-                        break;
-                    }
+                    let client = Client::new(
+                        self.main_loop_sender.clone(),
+                        self.connection.sender.clone(),
+                    );
+
                     let task = match msg {
                         Message::Request(req) => {
                             self.session
                                 .request_queue_mut()
                                 .incoming_mut()
                                 .register(req.id.clone(), req.method.clone());
 
+                            if self.session.is_shutdown_requested() {
+                                tracing::warn!(
+                                    "Received request after server shutdown was requested, discarding"
+                                );
+                                client.respond_err(
+                                    req.id,
+                                    lsp_server::ResponseError {
+                                        code: lsp_server::ErrorCode::InvalidRequest as i32,
+                                        message: "Shutdown already requested.".to_owned(),
+                                        data: None,
+                                    },
+                                )?;
+                                continue;
+                            }
+
                             api::request(req)
```

Comment:
Should we avoid adding the request in the queue if the server is responding right away?

Reply:
Actually, we need to register the request when using `client.respond` because it checks if the request is registered before responding. We could use `client_sender` directly to avoid the registration but I rather use the "proven" way and it's an edge case anyway, so that performance isn't that relevant.